collector-rabbitmq start fails when the queue exists

If rabbitmq queue exists and args of the queue not same as args of collector-rabbitmq default, zipkin server  would start fail .

exception file/line: 
---------------------
at zipkin.collector.rabbitmq.RabbitMQCollector$LazyInit.compute(RabbitMQCollector.java:141) ~[zipkin-collector-rabbitmq-2.8.4.jar!/:?]

stack trace:
---------------------
Caused by: com.rabbitmq.client.ShutdownSignalException: channel error; protocol method: #method<channel.close>(reply-code=406, reply-text=PRECONDITION_FAILED - inequivalent arg 'x-message-ttl'for queue 'queue.monitor.zipkin' in v
host 'monitor': received none but current is the value '36000000' of type 'long', class-id=50, method-id=10)
        at com.rabbitmq.utility.ValueOrException.getValue(ValueOrException.java:66) ~[amqp-client-5.1.2.jar!/:5.1.2]
        at com.rabbitmq.utility.BlockingValueOrException.uninterruptibleGetValue(BlockingValueOrException.java:36) ~[amqp-client-5.1.2.jar!/:5.1.2]
        at com.rabbitmq.client.impl.AMQChannel$BlockingRpcContinuation.getReply(AMQChannel.java:494) ~[amqp-client-5.1.2.jar!/:5.1.2]
        at com.rabbitmq.client.impl.AMQChannel.privateRpc(AMQChannel.java:288) ~[amqp-client-5.1.2.jar!/:5.1.2]
        at com.rabbitmq.client.impl.AMQChannel.exnWrappingRpc(AMQChannel.java:138) ~[amqp-client-5.1.2.jar!/:5.1.2]
        at com.rabbitmq.client.impl.ChannelN.queueDeclare(ChannelN.java:947) ~[amqp-client-5.1.2.jar!/:5.1.2]
        at com.rabbitmq.client.impl.recovery.AutorecoveringChannel.queueDeclare(AutorecoveringChannel.java:333) ~[amqp-client-5.1.2.jar!/:5.1.2]
        at zipkin.collector.rabbitmq.RabbitMQCollector$LazyInit.compute(RabbitMQCollector.java:141) ~[zipkin-collector-rabbitmq-2.8.4.jar!/:?]
        at zipkin.collector.rabbitmq.RabbitMQCollector$LazyInit.compute(RabbitMQCollector.java:128) ~[zipkin-collector-rabbitmq-2.8.4.jar!/:?]
        at zipkin.internal.LazyCloseable.tryCompute(LazyCloseable.java:51) ~[io.zipkin.java-zipkin-2.8.4.jar!/:?]
        at zipkin.internal.Lazy.get(Lazy.java:37) ~[io.zipkin.java-zipkin-2.8.4.jar!/:?]
        at zipkin.collector.rabbitmq.RabbitMQCollector.start(RabbitMQCollector.java:108) ~[zipkin-collector-rabbitmq-2.8.4.jar!/:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_72]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_72]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_72]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_72]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeCustomInitMethod(AbstractAutowireCapableBeanFactory.java:1831) ~[spring-beans-5.0.6.RELEASE.jar!/:5.0.6.RELEASE]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1774) ~[spring-beans-5.0.6.RELEASE.jar!/:5.0.6.RELEASE]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1702) ~[spring-beans-5.0.6.RELEASE.jar!/:5.0.6.RELEASE]
        ... 23 more

--- a/zipkin-collector/rabbitmq/src/main/java/zipkin2/collector/rabbitmq/RabbitMQCollector.java
+++ b/zipkin-collector/rabbitmq/src/main/java/zipkin2/collector/rabbitmq/RabbitMQCollector.java
@@ -168,7 +168,7 @@ public final class RabbitMQCollector extends CollectorComponent {
             (builder.addresses == null)
                 ? builder.connectionFactory.newConnection()
                 : builder.connectionFactory.newConnection(builder.addresses);
-        connection.createChannel().queueDeclare(builder.queue, true, false, false, null);
+        declareQueueIfMissing(connection);
       } catch (IOException | TimeoutException e) {
         throw new IllegalStateException("Unable to establish connection to RabbitMQ server", e);
       }
@@ -179,7 +179,7 @@ public final class RabbitMQCollector extends CollectorComponent {
         String name = RabbitMQSpanConsumer.class.getName() + i;
         try {
           // this sets up a channel for each consumer thread.
-          // We don't track channels them, as the connection will close its channels implicitly
+          // We don't track channels, as the connection will close its channels implicitly
           Channel channel = connection.createChannel();
           RabbitMQSpanConsumer consumer = new RabbitMQSpanConsumer(channel, collector, metrics);
           channel.basicConsume(builder.queue, true, name, consumer);
@@ -188,6 +188,23 @@ public final class RabbitMQCollector extends CollectorComponent {
         }
       }
       return connection;
+    }
+
+    private void declareQueueIfMissing(Connection connection) throws IOException, TimeoutException {
+      Channel channel = connection.createChannel();
+      try {
+        // check if queue already exists
+        channel.queueDeclarePassive(builder.queue);
+        channel.close();
+      } catch (IOException maybeQueueDoesNotExist) {
+        if (maybeQueueDoesNotExist.getCause() != null && maybeQueueDoesNotExist.getCause().getMessage().contains("NOT_FOUND")) {
+          channel = connection.createChannel();
+          channel.queueDeclare(builder.queue, true, false, false, null);
+          channel.close();
+        } else {
+          throw maybeQueueDoesNotExist;
+        }
+      }
     }
   }
 

--- a/zipkin-collector/rabbitmq/src/test/java/zipkin2/collector/rabbitmq/ITRabbitMQCollector.java
+++ b/zipkin-collector/rabbitmq/src/test/java/zipkin2/collector/rabbitmq/ITRabbitMQCollector.java
@@ -18,6 +18,8 @@ import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.concurrent.TimeoutException;
+
+import com.rabbitmq.client.Channel;
 import org.junit.After;
 import org.junit.ClassRule;
 import org.junit.Rule;
@@ -25,6 +27,8 @@ import org.junit.Test;
 import org.junit.rules.ExpectedException;
 import zipkin2.Span;
 import zipkin2.codec.SpanBytesEncoder;
+import zipkin2.collector.CollectorMetrics;
+import zipkin2.storage.InMemoryStorage;
 
 import static org.assertj.core.api.Java6Assertions.assertThat;
 import static zipkin2.TestObjects.LOTS_OF_SPANS;
@@ -113,6 +117,25 @@ public class ITRabbitMQCollector {
     Thread.sleep(1000);
     assertThat(rabbit.rabbitmqMetrics.messages()).isEqualTo(5);
     assertThat(rabbit.rabbitmqMetrics.messagesDropped()).isEqualTo(3);
+  }
+
+  /** See GitHub issue #2068 */
+  @Test
+  public void startsWhenConfiguredQueueAlreadyExists() throws IOException, TimeoutException {
+    Channel channel = rabbit.collector.connection.get().createChannel();
+    // make a queue with non-default properties
+    channel.queueDeclare("zipkin-test2", true, false, false, Collections.singletonMap("x-message-ttl", 36000000));
+    try {
+      RabbitMQCollector.builder()
+        .storage(InMemoryStorage.newBuilder().build())
+        .metrics(CollectorMetrics.NOOP_METRICS)
+        .queue("zipkin-test2")
+        .addresses(Collections.singletonList(rabbit.address())).build()
+        .start().close();
+    } finally {
+      channel.queueDelete("zipkin-test2");
+      channel.close();
+    }
   }
 
   /** Guards against errors that leak from storage, such as InvalidQueryException */

